### PR TITLE
Smoke test that runs in travis

### PR DESCRIPTION
**Pull request by KenCoder**
Don't merge yet - I intentionally put a console.error statement to see if this fails travis on the PR build.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -21,7 +21,7 @@
     "build-compare": "webpack --config ./compare/webpack.config.js",
     "dev-compare": "webpack-dev-server --content-base ./compare/output --config ./compare/webpack.config.js",
     "integration-test": "rm -rf newdir && mkdir newdir && cd newdir && node ../cli/index.js genConfig && node ../cli/index.js reference && node ../cli/index.js test && node -e \"require('../')('test')\"",
-    "smoke-test": "cd test/configs/; node ../../cli/index.js test --config=backstop_features",
+    "smoke-test": "node test/smoke/smoke_test.js",
```
Comment (garris) [inline package.json]: Very nice harness you built there!  

A suggestion: would you consider leaving the `smoke-test` command the same (so it executes the backstop_features test like before) -- and add a *new* command for your new harness.  Maybe call it like, `travis-test` or `git-commit-test` or `deploy-test`? 

I am totally open to suggestions -- my instinct is to keep it convenient for devs to still be able to run the raw smoke-test and see results (passing or non-passing). 

Does this make sense?

Comment (KenCoder) [reply package.json]: I think the harness works well for devs too because it tells them if it succeeds at the end. Right now it's hard to notice if something breaks (e.g., incorrect # of comparison failures or that a timeout). There are some scary looking messages in the output that I don't know if they mean somethings wrong or not (selector not found, etc).

The only difference is that the CI test doesn't pop up the browser report at the end. 

But I will add an argument so if you do
npm run smoke-test 

It will behave just like today with additional niceness and popup the browser, and 

npm run smoke-test -- ci

Will do it without the browser report.

Question for you: Will this have to run in Docker? It seems like launching in travis doesn't really work. If so I can't currently see the docker cloud config - maybe you can share it with me?

Comment (garris) [reply package.json]: Sounds good!

I honestly am not an expert in docker -- I just know enough to get me into trouble.  I am happy to share the configs with you -- they should all be in the `/docker` directory.

Also -- just so I am understanding correctly -- when you say...
> It seems like launching in travis doesn't really work
...are you saying you cant run backstop tests in travis?

Comment (KenCoder) [reply package.json]: > It seems like launching in travis doesn't really work

Well that's the way it is now - you can click below and see the error. I wasn't sure if you already knew this. If not - no worries I think I can fix this problem. So expect an updated PR shortly.

**Comment by KenCoder** [comment]
Don't merge yet! I'm still working on it - unfortunately chrome in Linux produces image mismatch with chrome on the mac, so likely the mac version will have to use a docker container to generate the reference bitmaps. I will let you know when it's ready, and no worries about work - I know how it is.

**Comment by garris** [comment]
😄 ok thank you.

BTW:  if you are running into rendering differences you could try adding this to the settings...
```
    "args": [
      "--no-sandbox",
      "--disable-setuid-sandbox",
      "--disable-gpu", 
      "--force-device-scale-factor=1", 
      "--disable-infobars=true"
    ]
```
In particular, —disable-gpu when used on Mac _and_ Linux may fix the font rendering issues.   It does seem to work across macOS boxes and _may_ work across Linux and Mac.

It would be good to know if this in fact can solve the issue.  I am happy to rerender using that setting if you don’t have a Mac handy.

And again, thanks so much for your effort 😊.

**Comment by KenCoder** [comment]
Hey - any luck reviewing this? Any concerns?

**Comment by garris** [comment]
Oh crap — sorry man.   So busy...   😿

I have a day to work on the library on 5/18 — but hopefully I can this  before that.

Thanks for your patience!

File: test/smoke/docker_smoke.sh [hunk 2]
```diff
@@ -0,0 +1,6 @@
+#!/usr/bin/env bash
```
Comment (garris) [inline test/smoke/docker_smoke.sh]: I would love to know what each of these lines did.  I find it hard to parse docker configs.   Particularly lines 5 & 6.   I guess I expect we would cd to a workspace, install backstop locally to that space and use the feature tests runner.   I don't understand why `/test/configs/linux/node_modules` is called out in there.

File: test/smoke/smoke_test.js [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+function sleep (ms) {
+  return new Promise(resolve => setTimeout(resolve, ms));
+}
+
+async function runSmokeTest (description, commands) {
+  const child = childProcess.spawn(commands[0], commands.slice(1),
+    { cwd: configsDir, env: processEnv }
+  );
+
+  const requiredStdout = [
```
Comment (garris) [inline test/smoke/smoke_test.js]: This sort of thing should be placed at the top of this script with explicit comments around how to use this array to handle different test scenarios.

In this case we should name it something very clear.

File: test/smoke/smoke_test.js [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    'Element not found for capturing: .monkey',
+    'report | 0 Failed'
+  ];
+
+  const allowedStdErr = [
```
Comment (garris) [inline test/smoke/smoke_test.js]: same comment as above for this array.

File: test/smoke/smoke_test.js [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    'Possible EventEmitter memory leak detected.'
+  ];
+
+  let missingStdOut = [...requiredStdout];
+
+  child.stdout.on('data', function (buf) {
```
Comment (garris) [inline test/smoke/smoke_test.js]: Please use named functions for clarity.

File: test/smoke/smoke_test.js [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    const line = String(buf);
+    util.print('StdErr:', line);
+    if (allowedStdErr.filter(allowed => line.includes(allowed)).length === 0) {
+      stdErrs.push(line.trim());
+    }
+  });
+
+  const code = await new Promise(resolve => child.on('close', resolve));
+
+  const problems = [];
```
Comment (garris) [inline test/smoke/smoke_test.js]: would prefer if these were encapsulated a bit by functions so that the overall flow of this script can be easily grasped in a few lines.

File: assets/styles.css [hunk 7]
```diff
@@ -43,7 +43,6 @@ a:hover {
   transform: translateZ(200px);
 }
 .lemurInYourFace {
-  transition: transform 750ms ease, opacity 750ms ease, visibility 800ms ease;
```
[the remark this answers is not in the document]
Comment (KenCoder) [reply assets/styles.css]: The way to do this is with the transitionend event. This could be a new backstop feature, or one can write custom Javascript to detect the transition and use the readyEvent feature of Backstop. There is already a test of readyEvent - this 'lemurInYourFace' is supposed to be the 'Simple' test case. Demonstrating CSS  transition in the "Simple" case seems wrong. 

How about we add another issue "demo CSS transition handling" that can be done in a separate PR? I didn't introduce this problem - I'm just cleaning up something that was already flaky.

[the remark this answers is not in the document]
Comment (garris) [reply assets/styles.css]: yes that works.  We can just comment the flaky test now and create an issue.

_There is actually an article on dwb  which discusses exactly your idea — using the transition event.  The onReady test script can attach a listener and just emit a console log which triggers the shot._

File: test/configs/backstop_features.js [hunk 8]
```diff
@@ -1,5 +1,8 @@
+const process = require('process');
+
 const ENGINE = 'puppet';
 const SCRIPT_PATH = 'puppet';
+const PLATFORM = process.env.BACKSTOP_SMOKE_PLATFORM || (process.platform === 'darwin' ? 'osx' : 'linux');
 
 module.exports = {
   id: `${ENGINE}_backstop_features`,
```
[the remark this answers is not in the document]
Comment (KenCoder) [reply test/configs/backstop_features.js]: Yes this is much nicer - I changed it to this.

File: test/configs/backstop_features.js [hunk 9]
```diff
@@ -83,7 +86,7 @@ module.exports = {
     },
     {
       label: 'cookies',
-      cookiePath: 'backstop_data/cookies.json',
+      cookiePath: `${PLATFORM}/backstop_data/cookies.json`,
```
[the remark this answers is not in the document]
Comment (KenCoder) [reply test/configs/backstop_features.js]: Fixed

File: test/configs/backstop_features.js [hunk 10]
```diff
@@ -123,17 +126,18 @@ module.exports = {
     }
   ],
   paths: {
-    bitmaps_reference: 'backstop_data/bitmaps_reference',
-    bitmaps_test: 'backstop_data/bitmaps_test',
-    engine_scripts: 'backstop_data/engine_scripts',
-    html_report: 'backstop_data/html_report',
-    ci_report: 'backstop_data/ci_report'
+    bitmaps_reference: `${PLATFORM}/backstop_data/bitmaps_reference`,
```
[the remark this answers is not in the document]
Comment (KenCoder) [reply test/configs/backstop_features.js]: Done

File: test/configs/backstop_features.js [hunk 11]
```diff
@@ -125,15 +128,16 @@ module.exports = {
   paths: {
     bitmaps_reference: 'backstop_data/bitmaps_reference',
     bitmaps_test: 'backstop_data/bitmaps_test',
-    engine_scripts: 'backstop_data/engine_scripts',
+    engine_scripts: 'engine_scripts',
```
Comment (garris) [inline test/configs/backstop_features.js]: Should also be under backstop data.

**Comment by gabegorelick** [comment]
Is there anything preventing this from being merged, aside from needing a rebase? More tests would give contributors a lot of confidence their changes won't break things.

**Comment by KenCoder** [comment]
We need to address #842 for the smoke tests to pass consistently. Then it needs to rebase.
